by-language/python-sqlalchemy/insert_pandas.py

```diff
@@ -0,0 +1,155 @@
+
+        # Create a DataFrame to feed into the database.
+        df = makeTimeDataFrame(nper=num_records, freq="S")
+        print(df)
+
+        logger.info(f"Connecting to {self.dburi}")
+        logger.info(f"Importing data with mode={mode}, bulk_size={bulk_size}, insertmanyvalues_page_size={insertmanyvalues_page_size}")
+
+        engine = self.get_engine(insertmanyvalues_page_size=insertmanyvalues_page_size)
+
+        # SQLAlchemy "Insert Many Values" mode. 40K records/s
+        # https://docs.sqlalchemy.org/en/20/core/connections.html#engine-insertmanyvalues
+        # https://docs.sqlalchemy.org/en/20/core/connections.html#engine-insertmanyvalues-page-size
+        if mode == "basic":
+            # Using `chunksize` does not make much of a difference here,
+            # because chunking will be done by SQLAlchemy already.
+            df.to_sql(name=self.table_name, con=engine, if_exists="replace", index=False)
+            # df.to_sql(name=self.table_name, con=engine, if_exists="replace", index=False, chunksize=bulk_size)
+
+        # Multi-row mode. It is slower.
+        # https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_sql.html
+        elif mode == "multi":
+            df.to_sql(name=self.table_name, con=engine, if_exists="replace", index=False, chunksize=bulk_size, method="multi")
+
+        # CrateDB bulk transfer mode. 65K records/s
+        # https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#bulk-operations
+        elif mode == "bulk":
+            df.to_sql(name=self.table_name, con=engine, if_exists="append", index=False, chunksize=bulk_size, method=insert_bulk)
+
+        else:
+            raise ValueError(f"Unknown mode: {mode}")
+
+    def show_table_stats(self):
+        """
+        Display number of records in table.
+        """
+        engine = self.get_engine()
+        with engine.connect() as conn:
+            conn.exec_driver_sql(f"REFRESH TABLE {self.table_name};")
+            result = conn.exec_driver_sql(f"SELECT COUNT(*) FROM {self.table_name};")
+            table_size = result.scalar_one()
+            logger.info(f"Table size: {table_size}")
+        #engine.dispose()
+
+
+def setup_logging(level=logging.INFO):
+    reset = escape_codes["reset"]
+    log_format = f"%(asctime)-15s [%(name)-26s] %(log_color)s%(levelname)-8s:{reset} %(message)s"
+
+    handler = colorlog.StreamHandler()
+    handler.setFormatter(colorlog.ColoredFormatter(log_format))
+
+    logging.basicConfig(format=log_format, level=level, handlers=[handler])
+
+    # Enable SQLAlchemy logging.
+    if SQLALCHEMY_LOGGING:
+        logging.getLogger("sqlalchemy").setLevel(level)
+
+
+@click.command()
+@click.option("--dburi", type=str, default="crate://localhost:4200", required=False, help="SQLAlchemy database connection URI.")
+@click.option("--mode", type=str, default="bulk", required=False, help="Insert mode. Choose one of basic, multi, bulk.")
```

Comment:
could use a choice option to make possible options strict.
```suggestion
@click.option("--mode", type=click.Choice(['basic', 'multi', 'bulk']), default="bulk", required=False, case_sensitive=False, help="Insert mode. Choose one of basic, multi, bulk.")
```

Reply:
Did you ask another expert on this?
```
TypeError: Parameter.__init__() got an unexpected keyword argument 'case_sensitive'
```